apps/api/src/app/widgets/usecases/mark-message-as-by-mark/mark-message-as-by-mark.usecase.ts

```diff
@@ -0,0 +1,116 @@
+
+import { MessageEntity, MessageRepository, SubscriberRepository, SubscriberEntity } from '@novu/dal';
+import { MarkMessagesAsEnum, WebSocketEventEnum } from '@novu/shared';
+import {
+  WebSocketsQueueService,
+  AnalyticsService,
+  InvalidateCacheService,
+  CachedEntity,
+  buildFeedKey,
+  buildMessageCountKey,
+  buildSubscriberKey,
+} from '@novu/application-generic';
+
+import { MarkMessageAsByMarkCommand } from './mark-message-as-by-mark.command';
+
+@Injectable()
+export class MarkMessageAsByMark {
+  constructor(
+    private invalidateCache: InvalidateCacheService,
+    private messageRepository: MessageRepository,
+    private webSocketsQueueService: WebSocketsQueueService,
+    private analyticsService: AnalyticsService,
+    private subscriberRepository: SubscriberRepository
+  ) {}
+
+  async execute(command: MarkMessageAsByMarkCommand): Promise<MessageEntity[]> {
+    await this.invalidateCache.invalidateQuery({
+      key: buildFeedKey().invalidate({
+        subscriberId: command.subscriberId,
+        _environmentId: command.environmentId,
+      }),
+    });
+
+    await this.invalidateCache.invalidateQuery({
+      key: buildMessageCountKey().invalidate({
+        subscriberId: command.subscriberId,
+        _environmentId: command.environmentId,
+      }),
+    });
+
+    const subscriber = await this.fetchSubscriber({
+      _environmentId: command.environmentId,
+      subscriberId: command.subscriberId,
+    });
+
+    if (!subscriber) throw new NotFoundException(`Subscriber ${command.subscriberId} not found`);
+
+    await this.messageRepository.changeMessagesStatus({
+      environmentId: command.environmentId,
+      subscriberId: subscriber._id,
+      messageIds: command.messageIds,
+      markAs: command.markAs,
+    });
+
+    const messages = await this.messageRepository.find({
+      _environmentId: command.environmentId,
+      _id: {
+        $in: command.messageIds,
+      },
+    });
+
+    await this.updateServices(command, subscriber, messages, command.markAs);
+
+    return messages;
+  }
+
+  private async updateServices(command: MarkMessageAsByMarkCommand, subscriber, messages, markAs: MarkMessagesAsEnum) {
+    this.updateSocketCount(subscriber, markAs);
+    const analyticMessage =
+      command.__source === 'notification_center'
+        ? `Mark as ${markAs} - [Notification Center]`
+        : `Mark as ${markAs} - [API]`;
+
+    for (const message of messages) {
+      this.analyticsService.mixpanelTrack(analyticMessage, '', {
+        _subscriber: message._subscriberId,
+        _organization: command.organizationId,
+        _template: message._templateId,
+      });
+    }
+  }
+
+  private updateSocketCount(subscriber: SubscriberEntity, markAs: MarkMessagesAsEnum) {
+    const eventMessage =
+      markAs === MarkMessagesAsEnum.READ || markAs === MarkMessagesAsEnum.UNREAD
+        ? WebSocketEventEnum.UNREAD
+        : WebSocketEventEnum.UNSEEN;
```

Comment:
maybe this should be moved to the helper function because it's duplicated in a couple of places, wdyt?

Reply:
Done :)